topdown/topdown_partial_test.go

```diff
@@ -3339,6 +3339,113 @@ func TestTopDownPartialEval(t *testing.T) {
 			}`},
 			wantQueries: []string{`"bar" = input.a; "baz" = input.b`},
 		},
+		{
+			note:    "ref heads: unknown rule value",
+			query:   "data.test.p.q[x]",
+			shallow: false,
+			modules: []string{`package test
+			p.q[x] := y {
+				x := "foo"
+				y := input.y
+			}`},
+			wantQueries: []string{`input.y; x = "foo"`},
+		},
+		{
+			note:  "ref heads: unknown ref var, unknown rule value",
+			query: "data.test.p.q[x]",
+			modules: []string{`package test
+			p.q[x] := y {
+				x := input.x
+				y := input.y
+			}`},
+			wantQueries: []string{`x = input.x; input.y`},
+		},
+		{
+			note:    "ref heads: unknown rule value, shallow inlining",
+			query:   "data.test.p.q.r[x]",
+			shallow: true,
+			modules: []string{`package test
+			p.q.r.s := y {
+				y := input.y
+			}`},
+			wantQueries: []string{`data.partial.test.p.q.r.s = x_term_0_0; x_term_0_0; x = "s"`},
+			wantSupport: []string{`package partial.test.p.q.r
+			s = __local0__1 { 
+				__local0__1 = input.y 
+			}`},
+		},
+		{
+			note:    "ref heads: unknown rule value, part-way query, shallow inlining",
+			query:   "y = data.test.p.q[x]",
+			shallow: true,
+			modules: []string{`package test
+			p.q.r.s := y {
+				y := input.y
+			}`},
+			wantQueries: []string{`data.test.p.q.r = y; x = "r"`},
```

Comment:
`data.test.p.q.r` will always be defined, yes. But even if an undefined `input.y` caused it to be undefined instead of `{}`, wouldn't this still be the expected optimized query?